### PR TITLE
Added explanation   poi frog

### PR DESCRIPTION
**Pull request by TheGamingMousse**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

NOTE: fast I/O is required for this problem (I'm pretty sure, as when I submitted w/o it TLEd)

### CODE REVIEW

File: solutions/platinum/poi-10-frog.mdx [hunk 1]
```diff
@@ -2,96 +2,163 @@
...
+we maintain a sliding window of size $k+1$. Essentially, we sweep
+left to right, and keep on moving the window right until either:
 
-typedef long long ll;
-typedef vector<int> vi;
-typedef pair<int, int> pii;
-template <class T>
-using Tree =
-    tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;
+$1.$ The window cannot be moved right any further.
```
Comment (ryanchou-dev) [inline solutions/platinum/poi-10-frog.mdx:13]:
```suggestion
1. The window cannot be moved right any further.
```
MDX has its own formatted lists

File: solutions/platinum/poi-10-frog.mdx [hunk 2]
```diff
@@ -2,96 +2,163 @@
...
 
-#define FOR(i, a, b) for (int i = a; i < (b); i++)
-#define F0R(i, a) for (int i = 0; i < (a); i++)
-#define FORd(i, a, b) for (int i = (b) - 1; i >= a; i--)
-#define F0Rd(i, a) for (int i = (a) - 1; i >= 0; i--)
+$2.$ Moving the window to the right would increase the maximum distance from this current location.
```
Comment (ryanchou-dev) [inline solutions/platinum/poi-10-frog.mdx:15]:
```suggestion
2. Moving the window to the right would increase the maximum distance from this current location.
```
^

File: solutions/platinum/poi-10-frog.mdx [hunk 3]
```diff
@@ -2,96 +2,163 @@
...
 
-int vis[MX];
+**Space Complexity:** $\mathcal{O}(N\log N)$
 
-void cycSize() {
-	FOR(i, 1, n + 1) {
-		int I = i;
-		for (; vis[I] == 0; I = nex[I][0]) vis[I] = i;
-		if (vis[I] != i) continue;
+## Implementation #1
```
Comment (ryanchou-dev) [inline solutions/platinum/poi-10-frog.mdx:37]: hmm, might be better to move these under the respective ## Solution headings?

**Review by ryanchou-dev — COMMENTED** [review]
code looks good!

File: solutions/platinum/poi-10-frog.mdx [hunk 4]
```diff
@@ -2,96 +2,163 @@
...
+	ll m;
+	cin >> n >> k >> m;
+	vector<ll> rocks(n);
+	for (ll &i : rocks) { cin >> i; }
+	vector<int> next_stone(n);
+	/*
+	 * Maintain a sliding window of
+	 * size k + 1 to calculate the next stone
+	 * the frog will jump to.
+	 */
```
Comment (SansPapyrus683) [inline solutions/platinum/poi-10-frog.mdx:55]:
```suggestion

// Maintain a sliding window of size k + 1 to
// calculate the next stone the frog will jump to.
```

File: solutions/platinum/poi-10-frog.mdx [hunk 5]
```diff
@@ -2,96 +2,163 @@
...
+	for (int i = 0; i < LOG; i++) {
+		if ((m >> i) & 1) {
+			for (int j = 0; j < n; j++) { res[j] = next_stone[res[j]]; }
 		}
+		vector<int> nxt_t(n);
+		for (int j = 0; j < n; j++) { nxt_t[j] = next_stone[next_stone[j]]; }
+		next_stone = nxt_t;
 	}
+	for (int i = 0; i < n; i++) { cout << res[i] + 1 << " \n"[i == n - 1]; }
+	return 0;
```
Comment (SansPapyrus683) [inline solutions/platinum/poi-10-frog.mdx:79]: no return 0

File: solutions/platinum/poi-10-frog.mdx [hunk 6]
```diff
@@ -2,96 +2,163 @@
...
 	cin >> n >> k >> m;
-	FOR(i, 1, n + 1) cin >> p[i];
-	genNex();
-	FOR(i, 1, 20) FOR(j, 1, n + 1) nex[j][i] = nex[nex[j][i - 1]][i - 1];
-	cycSize();
-	FOR(i, 1, n + 1) solve(i);
+	vector<ll> rocks(n);
+	for (ll &i : rocks) { cin >> i; }
+	const int LOG = 32 - __builtin_clz(n);
+	vector<vector<int>> lift(n, vector<int>(LOG));
```
Comment (SansPapyrus683) [inline solutions/platinum/poi-10-frog.mdx:97]: would recommend you just use std log functions: https://en.cppreference.com/w/cpp/numeric/math/log2

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
based

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
can u add language sections 🥺

File: solutions/platinum/poi-10-frog.mdx [hunk 7]
```diff
@@ -2,96 +2,171 @@
...
+	}
 
-void genNex() {
-	pii cur = {1, k + 1};
-	FOR(i, 1, n + 1) {
-		while (cur.s < n && p[cur.s + 1] - p[i] < p[i] - p[cur.f])
-			cur.f++, cur.s++;
-		if (p[i] - p[cur.f] >= p[cur.s] - p[i]) nex[i][0] = cur.f;
-		else nex[i][0] = cur.s;
+	const int LOG = (int)log2l(m) + 1;
```
Comment (SansPapyrus683) [inline solutions/platinum/poi-10-frog.mdx:57]: i mean u only use this once, why  not put it directly in the for loop with a <=

Comment (TheGamingMousse) [reply solutions/platinum/poi-10-frog.mdx:57]: true, lemme do that in a sec

**Review by SansPapyrus683 — DISMISSED** [review]
show SS of ac and we're chilling

File: solutions/platinum/poi-10-frog.mdx [hunk 8]
```diff
@@ -2,96 +2,171 @@
...
+
+1. The window cannot be moved right any further.
+2. Moving the window to the right would increase the maximum distance from this current location.
+
+Afterwards, we can proceed to binary lift to get our answer. However, directly calculating binary lifts
+the traditional way would take up too much memory. From here, there are two ways to proceed:
+
+## Implementation 1
+
+Iterate over the power of two we are lifting on in increasing order. Similar to binary exponentiation,
```
Comment (dongliu0426) [inline solutions/platinum/poi-10-frog.mdx:22]: might want to mention that a [linear memory binary lifting](https://codeforces.com/blog/entry/74847) exists so we can answer queries online

**Comment by SansPapyrus683** [comment]
@dongliu0426 saw you left a comment?

**Comment by TheGamingMousse** [comment]
I mentioned the linear memory binlift article @dongliu0426 linked.

File: solutions/platinum/poi-10-frog.mdx [hunk 9]
```diff
@@ -2,96 +2,175 @@
...
+	}
 }
 ```
+
+</CPPSection>
+</LanguageSection>
+
+Note: While not necessary for this problem, you can use
+[linear memory binary lifting](https://codeforces.com/blog/entry/74847)
+to answer the queries online.
```
Comment (SansPapyrus683) [inline solutions/platinum/poi-10-frog.mdx:176]: put this on top in like an <Info> tag

**Comment by TheGamingMousse** [comment]
idk if using the <Optional> component is the best... it just seemed the most fitting to me

**Review by SansPapyrus683 — APPROVED** [review]
W

**Comment by SansPapyrus683** [comment]
you should apply for content author :heart:

**Comment by TheGamingMousse** [comment]
> you should apply for content author ❤️

image

**Comment by dongliu0426** [comment]
> @dongliu0426 saw you left a comment?

right, I originally thought the graph was a tree, so I pointed out the linear memory binary lifting implementation. I realized that it doesn't quite work the same way for a functional graph, so I deleted it. I guess we can still convert it to a forest + cycles like the second solution does and apply the algorithm, but it's also possible to just DFS offline on the forest and answer queries in O(1) instead of doing any type of binary lifting.

**Pull request merged**